app/connectors/CodeListConnector.scala

```diff
@@ -42,62 +42,73 @@ object CodeItem {
+
+  type CodeMap[T <: CommonCode] = Map[String, ListMap[String, T]]
+
+  def getHolderOfAuthorisationCodes(locale: Locale): ListMap[String, HolderOfAuthorisationCode]
+  def getProcedureCodes(locale: Locale): ListMap[String, ProcedureCode]
+  def getProcedureCodesForC21(locale: Locale): ListMap[String, ProcedureCode]
+  def getAdditionalProcedureCodesMap(locale: Locale): ListMap[String, AdditionalProcedureCode]
+  def getAdditionalProcedureCodesMapForC21(locale: Locale): ListMap[String, AdditionalProcedureCode]
 
   val WELSH = new Locale("cy", "GB", "");
+  val supportedLanguages = Seq(ENGLISH, WELSH)
 }
 
 @Singleton
 class FileBasedCodeListConnector @Inject()(appConfig: AppConfig) extends CodeListConnector {
 
-  private implicit val supportedLanguages = Seq(ENGLISH, WELSH)
-
-  private def readCodesMapFromFile[T <: CommonCode](srcFile: String, factory: (CodeItem, Locale) => T) = {
+  private def loadCommonCodesAsOrderedMap[T <: CommonCode](srcFile: String, factory: (CodeItem, Locale) => T): CodeMap[T] = {
     val codeList = JsonFile.getJsonArrayFromFile(srcFile, CodeItem.formats)
 
     val langCodes = supportedLanguages.map { locale =>
-      val procedureCodes = codeList
+      val commonCodeList = codeList
         .map(factory(_, locale))
-        .map(cc => (cc.code, cc))
+        .map(commonCode => (commonCode.code, commonCode))
 
-      (locale.getLanguage -> ListMap(procedureCodes: _*))
+      (locale.getLanguage -> ListMap(commonCodeList: _*))
     }
 
     ListMap(langCodes: _*)
   }
 
-  private val procedureCodeListsByLang = readCodesMapFromFile(
+  private lazy val holderOfAuthorisationCodeListsByLang = loadCommonCodesAsOrderedMap(
+    appConfig.holderOfAuthorisationCodes,
+    (codeItem: CodeItem, locale: Locale) => HolderOfAuthorisationCode(codeItem.code, codeItem.getDescriptionByLocale(locale))
+  )
+
+  private lazy val procedureCodeListsByLang = loadCommonCodesAsOrderedMap(
     appConfig.procedureCodesListFile,
     (codeItem: CodeItem, locale: Locale) => ProcedureCode(codeItem.code, codeItem.getDescriptionByLocale(locale))
   )
 
-  private val procedureCodeForC21ListsByLang = readCodesMapFromFile(
+  private lazy val procedureCodeForC21ListsByLang = loadCommonCodesAsOrderedMap(
     appConfig.procedureCodesForC21ListFile,
     (codeItem: CodeItem, locale: Locale) => ProcedureCode(codeItem.code, codeItem.getDescriptionByLocale(locale))
   )
 
-  private val additionalProcedureCodeMapsByLang = readCodesMapFromFile(
+  private lazy val additionalProcedureCodeMapsByLang = loadCommonCodesAsOrderedMap(
     appConfig.additionalProcedureCodes,
     (codeItem: CodeItem, locale: Locale) => AdditionalProcedureCode(codeItem.code, codeItem.getDescriptionByLocale(locale))
   )
 
-  private val additionalProcedureCodeForC21MapsByLang = readCodesMapFromFile(
+  private lazy val additionalProcedureCodeForC21MapsByLang = loadCommonCodesAsOrderedMap(
```

Comment:
Don't use lazy keyword. These maps need to be loaded on app startup so that they fail fast if the data files (that get read to populate them) are not found for any reason.

Reply:
I'll remove the lazy but that's another reason why I strongly disagree with this implementation. I cannot test "Holder of Authorisation codes" in isolation because if the vals are not lazy they get all loaded when the class is instantiated.

I would go instead with different **small** classes, connectors or file loaders, called when needed by **UI** service classes for each object, because "Holder of Authorisation codes" are not "Procedure codes", and so on for all other "codes", even if they have the same structure.

I really dislike this idea of a class handling multiple business objects that have nothing in common but structure.
I think this `CommonCode` is a really bad idea, it places constraints and decreases flexibility.